### ISSUE TITLE
Snapshot to HDF5

### ISSUE THREAD

**Issue by Ethan-DeBandi99** [Bears-R-Us/arkouda#2512]
In order to allow users to quickly recover in the event of server crash or to resume work at a later date, we will be adding a snapshot feature. Initially, this will take all registered objects and write them to an HDF5 file (options to capture more objects will be added in a later update). A restore feature will be added along with this that will allow the user to restore these objects.

The initial restore implementation will return a dictionary of objects, but this will be further developed to recreate the Python variables on the client.

**Comment by Ethan-DeBandi99** [comment]
One thing that I just encountered that may be beneficial in this setting is to configure DataFrames to be saved under a group with a name and then have each column as a dataset.

I would worry that this will impact how we currently save dataframes and being able to read them in. Though I believe it would be possible. As it standas, dataframes would be reloaded, but as individual columns the dataframe would not be recreated.

**Comment by Ethan-DeBandi99** [comment]
I am thinking about how we want to name the snapshot files. My first thought was to just have the users provide a path to where they want the files saved and then the filename would be autogenerated using epoch time. However, I am thinking in some cases users may want to provide a custom name.

**Comment by Ethan-DeBandi99** [comment]
I have all the components configured for this. I am waiting for feedback from a few users I reached out to on handling dataframes. Once a decision is made there, I will wrap this up and post a PR.

**Comment by Ethan-DeBandi99** [comment]
Initially, we decided to add handling for DataFrames within HDF5 like we do for other complex objects. However, after working through some of the initial implementation I have identified some potential shortcomings with doing so. I have an alternate solution detailed below.

- Save DataFrame objects in their own file, these are indicated by the filename having `_DATAFRAME_#` appended to them.
- In these files, add a `AK_SNAPSHOT_META` group used to store an attribute containing the variable name for the DataFrame object
- The `AK_SNAPSHOT_META` would be ignored if the file is accessed through a standard read, but leveraged by the `restore` feature to recreate the appropriate variable.

This prevents issues with the variable name containing and underscore, but it will also prevent any changes from being required with existing DataFrame save/load workflows.

I am also seeing that we will need a `load_snapshot` function that loads a snapshot and returns a dictionary. This will be useful if a user is loading a snapshot, but is not aware of what the variable names where at the time of the snapshot. In this case, the dictionary keys would be the variable name and the values would be the object.
